dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+              bStride = 12;
+            } else if (bStride == 0 && a.type == "VEC2") {
+              // 2 components * 4 bytes
+              bStride = 8;
+            }
+            // we only use floats for vertex attributes
+            if (a.componentType != 5126)
+              count = 0;
+            if (count > (int)verts.size())
+              verts.resize(count);
+            // create a GLfloat array from the vertex attribute buffer offset
+            GLfloat *vBuf = reinterpret_cast<GLfloat *>(
+                &buffers[bv.buffer][bv.offset + a.offset]);
+            _VERBOUT_ << "Buffer: " << bv.buffer << std::endl;
+            _VERBOUT_ << "Byte offset: " << (bv.offset + a.offset) << std::endl;
+            _VERBOUT_ << "Byte stride: " << bStride << std::endl;
+            _VERBOUT_ << count << " vertex attribute vectors" << std::endl;
+            for (int i = 0; i < count; i++) {
+              GLfloat *vDat = &vBuf[bStride / 4 * i];
+              switch (atType) {
+              case 0:
+                verts[i].Position[0] = vDat[0];
+                verts[i].Position[1] = vDat[1];
+                verts[i].Position[2] = vDat[2];
+                break;
+              case 1:
+                verts[i].Normal[0] = vDat[0];
+                verts[i].Normal[1] = vDat[1];
+                verts[i].Normal[2] = vDat[2];
+                break;
+              case 2: // tangents are not supported by the phong_anim shader
+                //verts[i].Tangent[0] = vDat[0];
+                //verts[i].Tangent[1] = vDat[1];
+                //verts[i].Tangent[2] = vDat[2];
+                break;
+              case 3:
+                verts[i].UV[0] = vDat[0];
+                // OpenGL UVs are vertically flipped, glTF UVs work like Vulkan UVs
+                verts[i].UV[1] = vDat[1] * -1.0f + 1.0f;
+              }
+            }
+            _VERBOUT_ << std::endl;
+          }
+          //}else{
+          //	_ERROUT_ << "Primitive does not have vertex attributes." <<
+          //std::endl;
+          //}
+
+          // TODO: really need to verify this
+          // if( p["indices"].IsNumeric() ){
+          Accessor a = accessors[p->indices()];
+          int indType = a.componentType;
+          int numInds = a.count;
+          BufferView bv = bufferviews[a.bufferView];
+          int bStride = bv.stride;
+          // create an untyped char array from the index buffer offset, to be
+          // converted to a type
+          char *indBuf = &buffers[bv.buffer][bv.offset + a.offset];
+          for (int i = 0; i < numInds; i++) {
+            fgl::Index index;
+            // get a value based on the data type indType, the array index i,
+            // and the byte stride bStride
+            switch (indType) {
+            case 5120:
+              index = (fgl::Index)indBuf[(bStride || 1) * i];
+              break;
+            case 5121:
+              index = (fgl::Index)(reinterpret_cast<unsigned char *>(
```

Comment:
you use C-style casts elsewhere..

Reply:
One of those things I don't know how to fix.

Reply:
It looks largely correct, if a bit complicated. Index buffers don't have strides, so you should be able to index with just `[i]`. What exactly is not working here?
Again, you need to find the exact opengl constants for these numbers. e.g. 5121 is 0x1401 which you can find in e.g. https://www.khronos.org/registry/OpenGL/api/GLES2/gl2.h (same for all versions)

Reply:
I don't think this goes to the index buffer, I think it goes to the vertex buffer, and all the byte alignment stuff has been too much of a challenge to give too much thought.

Reply:
well, give it some more thought

Reply:
It's using GL constants now. As for what is or isn't working, I don't know.

Reply:
Well, I expect you to know before this gets merged :)

Reply:
Everything (including the `reinterpret_cast`s) was working when I used my own renderer. When I began to switch to Lobster's renderer, everything stopped rendering. It's probably not a mismatch in vertex attributes as I think I fixed that (unless there's a typo). It could be a mismatch in uniforms, a matrix math issue, some weird data alignment corruption issue (as I suspect based on the blue flickering when you fly to the floor level), or another GL error (easily tested with glGetError*). So I'll have to work on that, but I doubt the problem is in this section of code.